### ISSUE TITLE
DDLDatabase tripping on multiple pieces of a Heroku pg_dump

### ISSUE THREAD

**Issue by brmeyer** [jOOQ/jOOQ#8907]
Hey @lukaseder! It's been a while -- hope all is well.

Not at all urgent (can get around it), but throwing this in just as a heads up.

**I'M MORE THAN HAPPY TO HELP CONTRIBUTE! But, could use some pointers on where to start on these...**

### Expected behavior and actual behavior:

DDLDatabase is awesome, and in looking through past issues, it looks like the parsers continue to improve for postgres.

It appears to be tripping on a few different items thrown in by Heroku Postgres. I ran the following:

`pg_dump --host=HOST --port=5432 --username=USERNAME --password -O -x -s DBNAME > structure.sql`

Generally speaking, all of these errors can be worked around with different postgres settings, pg_dump switches, etc. But if it's at all possible to support or ignore what's being used, might be helpful since Heroku Postgres has become so bloody prevalent.

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

Here's the DDL, trimmed down to a single table:

```
--
-- PostgreSQL database dump
--

-- Dumped from database version 9.5.15
-- Dumped by pg_dump version 10.6 (Ubuntu 10.6-0ubuntu0.18.04.1)

SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET client_min_messages = warning;
SET row_security = off;

--
-- Name: plpgsql; Type: EXTENSION; Schema: -; Owner: -
--

CREATE EXTENSION IF NOT EXISTS plpgsql WITH SCHEMA pg_catalog;


--
-- Name: EXTENSION plpgsql; Type: COMMENT; Schema: -; Owner: -
--

COMMENT ON EXTENSION plpgsql IS 'PL/pgSQL procedural language';


SET default_tablespace = '';

SET default_with_oids = false;

--
-- Name: user_id_seq; Type: SEQUENCE; Schema: public; Owner: -
--

CREATE SEQUENCE public.user_id_seq
    START WITH 1
    INCREMENT BY 1
    NO MINVALUE
    NO MAXVALUE
    CACHE 1;


--
-- Name: users; Type: TABLE; Schema: public; Owner: -
--

CREATE TABLE public.users (
    id bigint NOT NULL,
    email character varying(255),
    first_name character varying(255),
    last_name character varying(255),
    phone character varying(255),
    password character varying(255),
    referral_code character varying(255),
    agreement_date timestamp without time zone,
    agreement_version character varying(255)
);


--
-- Name: users users_email_key; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.users
    ADD CONSTRAINT users_email_key UNIQUE (email);


--
-- Name: users users_pkey; Type: CONSTRAINT; Schema: public; Owner: -
--

ALTER TABLE ONLY public.users
    ADD CONSTRAINT users_pkey PRIMARY KEY (id);


--
-- PostgreSQL database dump complete
--
```

Here's what the parser trips on:

**SQL**

`SELECT pg_catalog.set_config('search_path', '', false);`

**ERROR**

```
org.jooq.impl.ParserException: Unknown function: [13:30] ...orming_strings = on;
SELECT pg_catalog.set_config([*]'search_path', '', false);
```

**SQL**

`CREATE EXTENSION IF NOT EXISTS plpgsql WITH SCHEMA pg_catalog;`

**ERROR**

```
org.jooq.impl.ParserException: GENERATOR, GLOBAL TEMPORARY TABLE, INDEX, OR ALTER VIEW, OR REPLACE VIEW, SCHEMA, SEQUENCE, TABLE, TEMPORARY TABLE, UNIQUE INDEX, or VIEW expected: [22:8] ...; Type: EXTENSION; Schema: -; Owner: -
--

CREATE [*]EXTENSION IF NOT EXISTS plpgsql WITH SCHEMA pg_catalog;
```

**SQL (#7751)**

`ALTER TABLE ONLY...`

**ERROR**

```
Error running jOOQ code generation tool: ADD, ALTER, COMMENT, DROP, MODIFY, or RENAME expected: [69:18] ...NT; Schema: public; Owner: -
[ERROR] --
[ERROR] 
[ERROR] ALTER TABLE ONLY [*]public.users
[ERROR]     ADD CONSTRAINT users_email_key UNIQUE (email);
```

### Versions:

- jOOQ: 3.12.0-SNAPSHOT
- Java: openjdk version "1.8.0_212"
- Database (include vendor): postgres 9.5.15

**Comment by lukaseder** [comment]
Thanks a lot for your thorough report, @brmeyer. All is well, I hope at your end too. Glad to see you're looking into using jOOQ again.

> I'M MORE THAN HAPPY TO HELP CONTRIBUTE! But, could use some pointers on where to start on these...

Our contribution guidelines are here: https://github.com/jOOQ/jOOQ/blob/master/CONTRIBUTING.md

In general, the answer to complex PRs is "no", by default :) - mostly because of the project management complexity third party PRs introduce. They usually solve problems very locally and specifically, when there are low hanging fruit for much better solutions, such as in the case of pgdump imports via `DDLDatabase`, there is the new "ignore comment syntax" in jOOQ 3.12, which allows for commenting certain SQL fragments only for jOOQ's parser, without any effect on the PostgreSQL parser: https://github.com/jOOQ/jOOQ/issues/8325

Regarding your specific issues:

> SELECT pg_catalog.set_config('search_path', '', false);

This will be addressed via https://github.com/jOOQ/jOOQ/issues/8678, which will be implemented as one of the last features in the upcoming jOOQ 3.12 (we'll hopefully be releasing early Q3 2019)

> CREATE EXTENSION IF NOT EXISTS plpgsql WITH SCHEMA pg_catalog;

This is such a common thing in PostgreSQL exports, I guess we should prioritise parsing and ignoring this statement: https://github.com/jOOQ/jOOQ/issues/8324. Of course, it can also be ignored using #8325 from jOOQ 3.12 onwards, but perhaps we can improve the UX of the `DDLDatabase` by recognising this common statement out of the box.

> ALTER TABLE ONLY...

Will look into this through https://github.com/jOOQ/jOOQ/issues/7751

So, with proper re-prioritisation, I guess this issue here can be closed as a duplicate of the other 3.

**Comment by lukaseder** [comment]
> SELECT pg_catalog.set_config('search_path', '', false);

In fact, parsing unknown functions (#8678) will not be sufficient to implement this feature, we'll have to tweak the `DDLDatabase` version of the parser to recognise this is the same thing as using `SET SCHEMA` as a syntax... I've created a new issue for this: #8910

**Comment by lukaseder** [comment]
All the issues mentioned here are now addressed for jOOQ 3.12.

**Comment by brmeyer** [comment]
Fantastic, thanks @lukaseder, will give the snapshot another...shot. Really appreciate the quick responses!

> In general, the answer to complex PRs is "no", by default :) - mostly because of the project management complexity third party PRs introduce. They usually solve problems very locally and specifically, when there are low hanging fruit for much better solutions...

Fully agree...
